test/boost/wasm_test.cc

```diff
@@ -0,0 +1,84 @@
+#include "lang/wasm.hh"
+#include "lang/wasm_instance_cache.hh"
+#include "rust/wasmtime_bindings.hh"
+#include "seastar/coroutine/maybe_yield.hh"
+#include <chrono>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/core/coroutine.hh>
+
+SEASTAR_TEST_CASE(test_long_udf_yields) {
+    auto wasm_engine = wasmtime::create_engine();
+    auto wasm_cache = std::make_unique<wasm::instance_cache>(100 * 1024 * 1024, std::chrono::seconds(1));
+    auto wasm_ctx = wasm::context(*wasm_engine, "fib", wasm_cache.get());
+    // Recursive fibonacci function
+    wasm::compile(wasm_ctx, {}, R"(
+(module
+  (type (;0;) (func (param i64) (result i64)))
+  (func (;0;) (type 0) (param i64) (result i64)
+    (local i64 i32)
+    local.get 0
+    i64.const 2
+    i64.lt_s
+    if  ;; label = @1
+      local.get 0
+      return
+    end
+    loop  ;; label = @1
+      local.get 0
+      i64.const 1
+      i64.sub
+      call 0
+      local.get 1
+      i64.add
+      local.set 1
+      local.get 0
+      i64.const 3
+      i64.gt_s
+      local.get 0
+      i64.const 2
+      i64.sub
+      local.set 0
+      br_if 0 (;@1;)
+    end
+    local.get 0
+    local.get 1
+    i64.add)
+  (func (;1;) (type 0) (param i64) (result i64)
+    local.get 0
+    call 0)
+  (memory (;0;) 2)
+  (global (;0;) i32 (i32.const 1024))
+  (export "memory" (memory 0))
+  (export "fib" (func 1))
+  (export "_scylla_abi" (global 0))
+  (data (;0;) (i32.const 1024) "01"))
+)");
+
+    auto argv = wasmtime::get_val_vec();
+    argv->push_i64(42);
+    auto rets = wasmtime::get_val_vec();
+    rets->push_i32(0);
+
+    auto store = wasmtime::create_store(wasm_ctx.engine_ptr);
+    auto instance = wasmtime::create_instance(wasm_ctx.engine_ptr, **wasm_ctx.module, *store);
+    auto func = wasmtime::create_func(*instance, *store, wasm_ctx.function_name);
+    auto memory = wasmtime::get_memory(*instance, *store);
+
+    auto fut = wasmtime::get_func_future(*store, *func, *argv, *rets);
+    bool stop = false;
+
+    using namespace std::chrono;
+    auto max_duration = duration(nanoseconds(0));
+    auto steps = 0;
+    while (!stop) {
+        auto start = system_clock::now();
+        stop = fut->resume();
+        max_duration = std::max(max_duration, system_clock::now() - start);
```

Comment:
This is likely to not work when running in parallel with other tests on overloaded machines.

Perhaps you can make it more robust by capturing CPU time (clock_gettime(CLOCK_THREAD_CPUTIME_ID)) and discarding measurements where the cpu time and real time differ by too much.

You can try to reproduce the problem by running many tight loops on the same cpu as the test, just to make sure you indeed fix the problem.

Reply:
Sounds good to me. I did manage to fail it using the tight loops approach and using the CLOCK_THREAD_CPUTIME_ID seems to help.

Reply:
> > I guess the separate stack is per-instance? Can it be moved to malloc too?
> 
> Yes, it's per instance and it's only using space when the instance is actually used (not while it's in the cache). There is currently no option to use stacks allocated using malloc, but it sounds like it would be a smaller problem to contribute to wasmtime with this change, than with the change of allocations of jit code

Eventually we'll want to contribute fragmented memory, where memory is an array of equally-sized blocks. It means loads/stores will translate to several machine instructions, but it does make grow O(1) (so it's useful not just to us with our fragmentation phobia, but to anyone who cares about latency).